packages/grid/x-data-grid-premium/src/components/GridGroupingCriteriaCell.tsx

```diff
@@ -62,6 +62,17 @@ export const GridGroupingCriteriaCell = (props: GridGroupingCriteriaCellProps) =
 
   const marginLeft = rootProps.rowGroupingColumnMode === 'multiple' ? 0 : rowNode.depth * 2;
 
+  let cellContent: React.ReactNode;
+
+  const colDef = apiRef.current.getColumn(rowNode.groupingField!);
+  if (typeof colDef.renderCell === 'function') {
+    cellContent = colDef.renderCell(props);
```

Comment:
I avoided calling custom `renderCell` functions inside another `renderCell` because users can have very customized cells with background on the whole cell, which would not go well on scenario like that.

This needs to be at least tested to see if the trade-off is correct

Reply:
> users can have very customized cells with background on the whole cell

I think it shouldn't be a problem. You can style `renderCell` differently if it's in a grouping row, or simply opt-out of rendering custom cell by doing:
```tsx
renderCell: (params) => {
  if (params.rowNode.type === 'group') {
    return params.value;
  }
}
```

Reply:
> users can have very customized cells with background on the whole cell

I've created this demo for testing it: https://codesandbox.io/s/magical-hill-68ons3?file=/demo.tsx
Currently, `height: 100%` in the grouping cell doesn't really make it full height. I can fix that in a separate PR.

Reply:
> This needs to be at least tested to see if the trade-off is correct

I've tested it with these custom components demos:
- https://codesandbox.io/s/tender-leavitt-by8js4?file=/demo.tsx
- https://codesandbox.io/s/friendly-carlos-jji9km?file=/demo.tsx